### PR TITLE
expression: refactor constant folding for IF&IFNULL

### PR DESCRIPTION
**Pull request by zz-jason**
Signed-off-by: Jian Zhang <zjsariel@gmail.com>

### What problem does this PR solve? 

1. If in strict sql mode, fold `ifnull` should not log `Data Truncated` warning messages
2. If in non-strict sql mode, fold `ifnull` should not return `Data Truncated` warnings to the client.

### What is changed and how it works?

1. Don't directly call `Constant.EvalInt()`, which results in error if the constant is a string. I changed it to call `Constant.Eval()` to get a `Datum`, use that `Datum` to check whether the first parameter is a `NULL` value.

2. If the first parameter for `IFNULL` is a constant and it's not a `NULL` value, we can directly return that constant parameter. In the old implementation, we returned the original `IFNULL` expression.

### Check List 

Tests 

 - Unit test ~TODO: add it later~ Done

Related changes

 - May need to cherry-pick it to the release branch, for it increases the usability of TiDB release versions. **Can be discussed later**.

### CODE REVIEW

File: expression/constant_fold.go [hunk 1]
```diff
@@ -59,16 +59,17 @@ func ifFoldHandler(expr *ScalarFunction) (Expression, bool) {
...
 	args := expr.GetArgs()
-	foldedArg0, _ := foldConstant(args[0])
+	foldedArg0, isDeferred := foldConstant(args[0])
 	if constArg, isConst := foldedArg0.(*Constant); isConst {
-		_, isNull0, err := constArg.EvalInt(expr.Function.getCtx(), chunk.Row{})
+		dt, err := constArg.Eval(chunk.Row{})
 		if err != nil {
 			log.Warnf("fold constant %s: %s", expr.ExplainInfo(), err.Error())
-			return expr, false
+			return constArg, isDeferred
```
Comment (eurekaka) [inline expression/constant_fold.go]: We should return `expr` instead of `constArg` here?

**Comment by zz-jason** [comment]
> > Don't directly call Constant.EvalInt(), which results in error if the constant is a string.
> 
> It seems that it's not correct to call `ToInt64` in Constant.EvalInt when a Datum is of type KindString since it goes against the original intent of this function.

If that datum is typed ENUM/BIT/SET, we have to call `Datum.ToInt64()` to get the correct value.

File: expression/constant_fold.go (hunk 1 above)
Comment (zz-jason) [reply expression/constant_fold.go]: Done.

**Comment by XuHuaiyu** [comment]
> > > Don't directly call Constant.EvalInt(), which results in error if the constant is a string.
> > 
> > It seems that it's not correct to call `ToInt64` in Constant.EvalInt when a Datum is of type KindString since it goes against the original intent of this function.
>
> If that datum is typed ENUM/BIT/SET, we have to call `Datum.ToInt64()` to get the correct value.
``` go
if c.GetType().Hybrid() || c.Value.Kind() == types.KindBinaryLiteral || c.Value.Kind() == types.KindString {}

func (ft *FieldType) Hybrid() bool {
	return ft.Tp == mysql.TypeEnum || ft.Tp == mysql.TypeBit || ft.Tp == mysql.TypeSet
}
```

`c.GetType.Hybrid()` already checked the it.

File: expression/constant_fold_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	tk.MustExec(`drop table if exists t;`)
+	tk.MustExec(`create table t(a bigint, b bigint);`)
+	tk.MustExec(`insert into t values(1, 1);`)
+	tk.MustQuery(`desc select ifnull("aaaa", a) from t;`).Check(testkit.Rows(
+		`Projection_3 10000.00 root "aaaa"`,
+		`└─TableReader_5 10000.00 root data:TableScan_4`,
+		`  └─TableScan_4 10000.00 cop table:t, range:[-inf,+inf], keep order:false, stats:pseudo`,
+	))
+	tk.MustQuery(`show warnings;`).Check(testkit.Rows())
+	tk.MustQuery(`select ifnull("aaaa", a) from t;`).Check(testkit.Rows("aaaa"))
```
Comment (XuHuaiyu) [inline expression/constant_fold_test.go]: Add a test case for `select ifnull(1=null, a) from t;`

Comment (zz-jason) [reply expression/constant_fold_test.go]: the purpose of this test is?

Comment (XuHuaiyu) [reply expression/constant_fold_test.go]: It seems it is not really needed.
Never mind.

**Pull request merged**